Glob performance

eatkins opened the pull request:
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) guidelines

This PR is built on top of #4535, #4537 and https://github.com/sbt/io/pull/216. I will update the PR as those other PRs are merged and a new io version is released.

High level, this replaces a number of file apis within sbt with glob based apis. This is good for performance because the glob based file look ups are generally routed through the file system cache. Even when the cache is not used, the performance should be good because of optimizations that I made in swoval for recursively listing a directory. Another benefit is that in watch, I can swap out the global file tree repository to determine which globs are actually used by a task. This enables watch to work even when the watch sources are unspecified or hidden from the dependency graph in a dynamic task.

I also actually implemented #3678, which allows clean to be scoped to the compile task and Compile / Test configurations. I copied some of the scripted tests from that #3678 and added a few additional scripted tests.

eatkins committed:
Add dsl for collecting globs

Right now, the sbt.internal.io.Source is something of a second class
citizen within sbt. Since sbt 0.13, there have been extension classes
defined that can convert a file to a PathFinder but no analog has been
introduced for sbt.internal.io.Source.

Given that sbt.internal.io.Source was not really intended to be part of
the public api (just look at its package), I think it makes sense to
just replace it with Glob. In this commit, I add extension
methods to Glob and Seq[Glob] that make it possible to easily
retrieve all of the files for a particular Glob within a task. The
upshot is that where previously, we'd have had to write something like:

watchSources += Source(baseDirectory.value / "src" / "main" / "proto", "*.proto", NothingFilter)

now we can write

watchGlobs += baseDirectory.value / "src" / "main" / "proto" * "*.proto"

Moreover, within a task, we can now do something like:
foo := {
  val allWatchGlobs: Seq[File] = watchGlobs.value.all
  println(allWatchSources.mkString("all watch source files:\n", "\n", ""))
}
Before we would have had to manually retrieve the files.

The implementation of the dsl uses the new GlobExtractor class which
proxies file look ups through a FileTree.Repository. This makes it so
that, by default, all file i/o using Sources will use the default
FileTree.Repository. The default is a macro that returns
`sbt.Keys.fileTreeRepository.value: @sbtUnchecked`. By doing it this
way, the default repository can only be used within a task definition
(since it delegates to `fileTreeRepository.value`). It does not,
however, prevent the user from explicitly providing a
FileTree.Repository instance which the user is free to instantiate
however they wish.

Bonus: optimize imports in Def.scala and Defaults.scala

eatkins committed:
Bump io

This new version of io breaks source and binary compatibility everywhere
that uses the register(path: Path, depth: Int) method that is defined on
a few interfaces because I changed the signature to register(glob:
Glob). I had to convert to using a glob everywhere that register was
called.

I also noticed a number of places where we were calling .asFile on a
file. This is redundant because asFile is an extension method on File
that just returns the underlying file.

Finally, I share the IOSyntax trait from io in AllSyntax. There was more
or less a TODO suggesting this change. The one hairy part is the
existence of the Alternative class. This class has unfortunately somehow
made it into the sbt package object. While I doubt many plugins are
using this, it doesn't seem worth breaking binary compatibility to get
rid of it. The issue is that while Alternative is defined private[sbt],
the alternative method in IOSyntax is public, so I can't get rid of
Alternative without breaking binary compatibility.

I'm not deprecating Alternative for now because the sbtProj still has
xfatal warnings on. I think in many, if not most, cases, the Alternative
class makes the code more confusing as is often the case with custom
operators. The confusion is mitigated if the abstraction is used only in
the file in which it's defined.

eatkins committed:
Update clean task to use globs

The clean task is unreasonably slow because it does a lot of redundant
io. In this commit, I update clean to be implemented using globs. This
allows us to (optionally) route io through the file system cache. There
is a significant performance improvement to this change. Currently,
running clean on the sbt project takes O(6 seconds) on my machine. After
this change, it takes O(1 second).

To implement this, I added a new setting cleanKeepGlobs to replace
cleanKeepFiles. I don't think that cleanKeepFiles returning Seq[File] is
a big deal for performance because, by default, it just contains the
history file so there isn't much benefit to accessing a single file
through the cache. The reason I added the setting was more for
consistency and to help push people towards globs in their own task
implementations.

Part of the performance improvement comes from inverting the problem.
Before we would walk the file system tree from the base and recursively
delete leafs and nodes in a depth first traversal. Now we collect all of
the files that we are interested in deleting in advance. We then sort
the results lexically by path name and then perform the deletions in
that order. Because children will always comes first in this scheme,
this will generally allow us to delete a directory.

There is an edge case that if files are created in a subdirectory after
we've created the list to delete, but before the subdirectory is
deleted, then that subdirectory will not be deleted. In general, this
will tend to impact target/streams because writes occur to
target/streams during traversal. I don't think this really matters for
most users. If the target directory is being concurrently modified with
clean, then the user is doing something wrong.

To ensure legacy compatibility, I re-implement cleanKeepFiles to return
no files. Any plugin that was appending files to the cleanKeepFiles task
with `+=` or `++=` will continue working as before because I explicitly
add those files to the list to delete. I updated the actions/clean-keep
scripted test to use both cleanKeepFiles and cleanKeepGlobs to ensure
both tasks are correctly used.

Bonus: add debug logging of all deleted files

eatkins committed:
Stop using ---/pair in SbtLauncherPlugin

I'd like to remove '---' and 'pair' in sbt 2 so I'm inlining the logic where
I find it. The '---' method is trivially implemented with a filter on
the sequence of files and filtering the output will not require io,
unlike '---'. For pair, I get confused every time I see it in the code
and it is rarely saving more than a line. While I understand that it may
have been convenient when the code using pair was originally written, I
don't think it is worth the maintenance cost. My specific issue is that
to me pair means tuple2, full stop. The definition of pair is:
def pair[T](mapper: File => Option[T], errorIfNone: Boolean = true): Seq[(File, T)]
First of all, it's not at all obvious when seen inline in the code that
it has the side effect of evaluating PathFinder.get. Moreover, it
doesn't return a general pair, it's a very specific pair with a File in
the first position. I just don't see how using pair improves upon, say:

val func: File => Option[(File, String)] = ???
globs.all.flatMap(func)

or

val func: File => Option[(File, String)] = ???
globs.all.map(f => func(f) match {
  case Some(r) => r
  case None => throw new IllegalStateException("Couldn't evaluate func for $f")
}) // or just define `func = File => (File, String)` and throw on an error

eatkins committed:
Rework FileTreeRepository configuration

The FileTreeViewConfig abstraction that I added was somewhat unwieldy
and confusing. The original intention was to provide users with a lot of
flexibility in configuring the global file tree repository used by sbt.
I don't think that flexibility is necessary and it was both conceptually
complicated and made the implementation complex. In this commit, I add a
new boolean flag enableGlobalCachingFileTreeRepository that toggles
which kind of FileTreeRepository to use globally.

There are actually three kinds of repositories that could be returned:
1) FileTreeRepository.default -- this caches the entire file system
   tree it hooks into the cache's event callbacks to create a file event
   monitor. It will be used if enableGlobalCachingFileTreeRepository is
   true and Global / pollingGlobs := Nil
2) FileTreeRepository.hybrid -- similar to FileTreeRepository.default
   except that it will not cache any files that are included in
   Global / pollingGlobs. It will be used if
   enableGlobalCachingFileTreeRepository is true and
   Global / pollingGlobs is non empty
3) FileTreeRepository.legacy -- does not cache any of the file system
   tree, but does maintain a persistent file monitoring process that is
   implemented with a WatchServiceBackedObservable. Because it doesn't
   poll, in general, it's ok to leave the monitoring on in the
   background. One reason to use this is that if there are any issues
   with the cache being unable to accurately mirror the underlying file
   system tree, this repository will always poll the file system
   whenever sbt requests the entries for a given glob. Moreover, the
   file system tree implementation is very similar to the implementation
   that was used in 1.2.x so this gives users a way to almost fully opt
   back in to the old behavior.

eatkins committed:
Avoid deprecated Alternative class

eatkins committed:
Support custom clean tasks

This rewroks the cleanTask so that it only removes a subset of the
files in the target directory. To do this, I add a new task, outputs,
that returns the glob representation of the possible output files for
the task. It must be a task because some outputs will depend on streams.
For each project, the default outputs are all of the files in
baseDirectory / target.

Long term, we could enhance the clean task to be automatically generated
in any scope (as an input task). We could then add the option for the
task scoped clean to delete all of the transitive outputs of the class.
That is beyond the scope of this commit, however.

I copied the scripted tests from #3678 and added an additional test to
make sure that the manage source directory was explicitly cleaned.

eatkins committed:
Move clean task implementation to Clean.scala

The Defaults.scala file has a lot going on. I am trying to generally
follow the pattern of implementing the default task implementation in a
different file and just adding the appropriate declarations in
Defaults.scala.

eed3si9n commented on sbt/src/sbt-test/actions/compile-clean/test:
```diff
@@ -0,0 +1,22 @@
+$ exists target/scala-2.12/classes/A.class
+$ exists target/scala-2.12/test-classes/B.class
+
+> test:compile::clean
```
I think this should be `Test/clean`. In other words, `clean` should just be scoped to `Test` configuration.

eatkins replied:
FWIW, I copied this test exactly from #3678. I think there is a valid use case for scoping both to the configuration and to the task separately. I might want to clear out the output files for just a specific task in a configuration but when i clean the configuration, I want to clear out the output files for all tasks in the configuration. Does that make sense? I don't mind dropping the task specific clean, but that was the motivation.

eed3si9n replied:
I think task scoping here is confusing since all sorts of tasks can trigger a compilation, including `test`.

eatkins replied:
To play devil's advocate, what if I just want to remove the jars generated by `package`? What if I want to remove the class files but not the jars? For me, it's more confusing if clean cannot be scoped at the task level. Most users are just going to use clean at the config level but I don't see why we need to exclude it at the task level.

eed3si9n replied:
One of the reasons is that I think task-scoped tasks are confusing concepts to users, and I want to avoid it as much as possible.

eatkins replied:
I get that, but the task scoped task is only relevant for advanced users who even know that syntax. As is, the config scoped clean works as expected and most users would never even think to try the task scoped version. Personally I think the real problem is the legacy `:` based syntax. When using the slash syntax, I don't think it's that confusing. Would it bother you less if the test was written `Test/compile/clean`? That works.

eatkins replied:
Oh, ugh. I see what you're saying a bit more now. Compile is really a special case because of the default configurations. I think my argument is more for arbitrary tasks, but there is all kinds of special casing around the compile task. That being said, I still think there is an argument that `compile/clean` should only delete the class files while `Compile/clean` would delete everything generated by the configuration.

eed3si9n replied:
When I type `clean`, what I often want is "clear the caching behavior and start from scratch", which doesn't necessarily have to do with wiping out the `target/` directory. Having that work for one task is tricky, let alone making sure it works consistently for all tasks transitively. Config-scoped clean deleting `classes/` is much more low-tech, but addresses a use case in a consistent way.

eatkins replied:
I disagree but I don't care enough to continue the debate so I'll just make it work for just the configuration.

eatkins replied:
https://github.com/sbt/sbt/compare/2da32eccd6f3e242d2859fa546d76f805b55a17d..f26afe6681da174e80e2ca6db8e638e1abf732cc

eatkins commented:
Sigh, windows is proving annoying wrt the clean task. Hopefully I'll have everything working by tomorrow afternoon.

eatkins committed:
Don't try to stamp files that don't exist

This was causing slowdowns in windows.

eatkins committed:
Run windows tests on visual studio 2015 and 2017

I've noticed that it's possible to write code that works on one but not
the other.

eatkins committed:
Reimplement clean

I ran into a couple of issues with the clean implementation. I changed
the logging to print to stdout instead of streams if enabled. I also
added a helper, Clean.deleteContents that recursively deletes all of the
contents of a directory except for those that match the exclude filter
parameter.

Using a normal logger was a bad idea because we are actually deleting
the target/streams directory when running clean.

The previous implementation worked by getting the full list of files to
delete, reverse sorting it and then deleting every element in the list.
While this can work well it many circumstances, if the directory is
still being written to during the recursive deletion, then we could miss
files that were added after we fetched all of the files. The new version
lazily lists the subdirectories as needed.

eatkins committed:
Fix repository setup on windows

Windows io really doesn't handle concurrent readers and writers all that
well. Using the LegacyFileTreeRepository was problematic in windows
scripted tests because even though the repository implementation did not
use the cache in its list methods, it did persistently monitor the
directories that were registered. The monitor has to do a lot of
io on a background thread to maintain the cache. This caused io
contention that would cause IO.createDirectory to fail with an obscure
AccessDeniedException. The way to avoid this is to prevent the
background io from occurring at all.

I don't necessarily think this will impact most users running sbt
interactively with a cache, but it did cause scripted tests to fail. For
that reason I made the default in non-interactive/shell use cases on
windows to be a PollingFileRepository which never monitors the file
system except when we are in a watch. The LegacyFileTreeRepository works
fine on mac and linux which have a more forgiving file system.

To make this work, I had to add FileManagement.toMonitoringRepository.
There are now two kinds of repositories that cannot monitor on their
own: HybridPollingFileTreeRepository and PollingFileRepository. The
FileManagement.toMonitoringRepository makes a new repository that turns
on monitoring for those two repository types and disables the close
method on all other repositories so that closing the FileEventMonitor
does not actually close the global file repository.

eatkins commented:
~~It turns out that the java watchservice for windows is not ideal for the sbt use case. It reports way too many events. For example, if you open a file, an event is created. I need to add a native windows watch service (similar to the one i wrote for macos) that filters watch events to swoval. This shouldn't be too hard because the two apis are similar (inotify is completely different). Also, I already have the test suite written.~~ I'm less certain that I need a different watch service. It could be something else in swoval.

eatkins commented:
Yeah, there was a nasty performance bug in swoval: https://github.com/sbt/io/pull/222

eatkins committed:
Fix equals for FileCacheEntry

The equals method didn't work exactly the way I thought. By delegating
to the equivStamp object in sbt we can be more confident that it is
actually comparing the stamp values and not object references or
some other equals implementation.

eatkins committed:
Add sbt io to the sbt project

Without this, the sbt io version is used by the compiler which means
that apis added in later versions of io are not available. I don't
understand why the transitive dependency on io is not used, but this
fixes the issue.

eatkins committed:
Make file repository apis private[sbt]

I haven't fully settled on these interfaces yet so they shouldn't be publicly
exposed in the sbt api.

eatkins committed:
Add Event trait to FileCacheEntry

Rather than exposing the FileEventMonitor.Event types, which are under
active development in the io repo, I am adding a new event trait to
FileCacheEntry. This trait doesn't expose any internal implementation
details.

eatkins committed:
Rename FileCacheEntry to FileAttributes

I decided that FileCacheEntry was a bad name because the methods did not
necessarily have anything to do with caching. Moreover, because it is
exposed in a public interface, it shouldn't be in the internal package.

eatkins committed:
Make Stamped.file private[sbt]

eatkins committed:
Don't expose TypedPath to users

I've decided I don't like the TypedPath interface so I'm not going to
expose it publicly.

eatkins committed:
Return (Path, FileAttributes) instead of Stamped.File

I realized that Stamped.File was a bad interface that was really just an
implementation detail of external hooks. I updated the
GlobLister.{ all, unique } methods to return Seq[(Path, FileAttributes)]
rather than Stamped.File which is a much more natural api and one I
could see surviving the switch to nio based apis planned for
1.4.0/2.0.0. I also added a simple scripted test for glob listing. The
GlobLister.all method is implicitly tested all over the place since the
compile task uses it, but it's good to have an explicit test.

eatkins committed:
Do not use caching repository by default

The caching repository does not work universally so set the default to
always poll. This is still faster than in sbt 1.2.x because of
performance improvements that I added for listing directories.

eatkins committed:
Temporarily make FileTree.Repository public

This should not be exposed to users but will be for M1.

eatkins commented:
I turned off the cache in this PR. I'll make it configurable, but not default, in #4512.

eed3si9n commented:
@eatkins Does this work for you locally?

```
[error] x source-dependencies/relative-source-error 
[error]  Cause of test exception: {line 3}  Command succeeded but failure was expected
```

eatkins committed:
Remove relative-source-error test

The io library now delegates to swoval for io which generally handles
relative sources correctly (by converting them to absolute paths
internally). This test started failing after the io bump because of
this. It seemed to me that relative sources not working was not a
feature so I just removed the test.

eatkins commented:
https://github.com/sbt/sbt/pull/4539/commits/ffa69ea5d690b5d8d548e0eefd98e83587a30de2

The commit message is arguably a little bit glib. Relative source directories are a bit tricky since sbt doesn't have a notion of current working directory so relative sources are implicitly relative to the sbt base directory. I guess marking the test PENDING is an option rather than deleting it, but it's not straightforward to make the previous semantics work with the new version of the swoval library (which automatically converts non relative paths to absolute paths internally at interface boundaries).

eed3si9n commented:
As long as https://github.com/sbt/sbt/issues/923 remains fixed, we should be ok.

eed3si9n reviewed (approved):
LGTM pending Travis
